README.md
Pbase
=====

Manipulating and exploring protein and proteomics data.

## Installation


From github using `devtools::install_github`:

	library("devtools")
	install_github("ComputationalProteomicsUnit/Pbase")

### Dependencies

See the `DESCRIPTION` file for a complete list.

## Getting started

Currently, the best way to get started is `?Proteins` and the
[`Pbase-data`](http://bioconductor.org/packages/devel/bioc/vignettes/Pbase/inst/doc/Pbase-data.html)
vignette. More documentation is on its way.

## Development

`Pbase` is under heavy development and is likely to considerably
change in the near future. Suggestion and bug reports are welcome and
can be filed as
[github issues](https://github.com/ComputationalProteomicsUnit/pbase/issues).

If you would like to contribute, please directly send pull requests
for minor contributions and typos. For major contributions, we suggest
to first get in touch with the package maintainers.

## Ideas

### Assessing the redundancy of a protein fasta database

Given a protein fasta file, what is the maximal sensitivity that can
be expected from a mass spectrometry experiment with 0, 1,
... miscleavages. This should probably also include a filtering step
for peptide *flyability*.

#### Flyability/Detectability

Some literature about estimating detectability:

- LogR: [Liu, Hui, et al. "The Prediction of Peptide Detectability in MS Data Analysis Using Logistic Regression." Bioinformatics and Biomedical Engineering,(iCBBE) 2011 5th International Conference on. IEEE, 2011.](http://dx.doi.org/10.1109/icbbe.2011.5780167)
- SVM: [Webb-Robertson, Bobbie-Jo M., et al. "A support vector machine model for the prediction of proteotypic peptides for accurate mass and time proteomics." Bioinformatics 24.13 (2008): 1503-1509.](http://dx.doi.org/10.1093/bioinformatics/btn218)
- NN: [Sanders, William S., et al. "Prediction of peptides observable by mass spectrometry applied at the experimental set level." BMC bioinformatics 8.Suppl 7 (2007): S23.](http://dx.doi.org/10.1186/1471-2105-8-S7-S23)
- Gausian Mixed Discrimination [Mallick, Parag, et al. "Computational prediction of proteotypic peptides for quantitative proteomics." Nature biotechnology 25.1 (2007): 125-131.](http://dx.doi.org/10.1038/nbt1275)

##### Liu et al. 2011:

Requirements for in-silico created peptides: `missedCleavages = 0:2`, `length(peptides) >= 6`, `mass(peptides) < 6000` (Da)

Logistic Regression based on Hydrophobicity, Isoelectric point, length,
molecular weight, average hydrophobicity, average isoelectric point

##### Webb-Robertson et al. 2007:

Requirements for in-silico created peptides: `missedCleavages = 0:2`, `length(peptides) >= 6`, `mass(peptides) < 6000` (Da)

35 features: length, weidght, # of (non-)polar, # of (un)charged, # of pos./neg. charged residues, hydrophobicity (different models), polarity (different models), bulkiness, AA singlet counts

##### Sanders et al. 2007

Requirements for in-silico created peptides: `length(peptides) >= 6`

Features: Length, Charge, Isoelectric Point, Molecular Weight, Hydropathicity, Counts of each AA (20 Features), Percent composition of each AA (20 Features), Percent of polar, psoitive, negative, hydrophobic AA

take-home-message: a model of one species/dataset could not be transfered to another dataset (without dramatically decreasing the performance)

##### Mallick et al. 2007

~1000 Features.

Some of the most discriminating properties:
Total/Average net/positive charge, hydrophobic moment, isoelectric point, Histidine composition

take-home-message: The model of one species is comparable to another if the evolutionary
distance is small (e.g. yeast and human) but you can't compare different devices/datasets (e.g. MALDI vs ESI)

###### Simple Rules

Mass: `500:4500`

http://www.nature.com/nbt/journal/v25/n1/extref/nbt1275-S5.pdf
http://ieeexplore.ieee.org/ielx5/5779756/5779971/5780167/html/img/5780167-fig-1-large.gif

Length: `5:40`

http://www.nature.com/nbt/journal/v25/n1/extref/nbt1275-S6.pdf
http://ieeexplore.ieee.org/ielx5/5779756/5779971/5780167/html/img/5780167-fig-1-large.gif

95% of all peptides are of length `5:30`:
http://www.nature.com/nbt/journal/v25/n1/extref/nbt1275-S24.pdf

Average Isoelectric point: `seq(0, 1.4)`
http://ieeexplore.ieee.org/ielx5/5779756/5779971/5780167/html/img/5780167-fig-1-large.gif

### Hydropathy/Hydrophobicity

http://web.expasy.org/tools/protparam/protparam-doc.html
http://web.expasy.org/compute_pi/pi_tool-doc.html
[Kyte, Jack, and Russell F. Doolittle. "A simple method for displaying the hydropathic character of a protein." Journal of molecular biology 157.1 (1982): 105-132.](http://dx.doi.org/10.1016/0022-2836(82)90515-0)

### Selection of optimal heavy peptides for absolute quantitation

See Pavel's [idea](https://github.com/sgibb/cleaver/issues/5).

### Protein domains

Available through the integration with the `EnsmbleDb` package. See the `Pbase-with-ensembldb` vignette.

### Mapping a Protein Sequence to a Genome Sequence

See the [`mapping`](http://bioconductor.org/packages/devel/bioc/vignettes/Pbase/inst/doc/mapping.html) vignette.

See also
[this document](https://github.com/ComputationalProteomicsUnit/Intro-Integ-Omics-Prot/blob/master/mapping.md)
for additional examples and integration with RNA-seq data.

## Interoperability

The package allows to easily interact with `AAString` and
`AAStringSet` instances, protein databases such as UniProt (and
possibly biomaRt in the future) using protein identifiers, protein
identification results (`mzID` or (devel) `mzR` packages) and possibly
also `MSnExp` and `MSnSet` instances.


NEWS
CHANGES IN VERSION 0.25.1
-------------------------
 o Removing mapping functions, moved to the ensembldb package. <2019-08-08 Thu>
 o Marking package for deprecation <2019-08-08 Thu>

CHANGES IN VERSION 0.15.3
-------------------------
 o Working up new API (see issue #41) <2017-04-09 Sun>
 o Temporarily remove some vignettes, until the API has
   stabilised. <2017-04-09 Sun>

CHANGES IN VERSION 0.15.2
-------------------------
 o New Proteins class implementation - see issue #38. <2016-12-10 Sat>
 o Fixed many errors from current rewrite <2017-04-08 Sat>

CHANGES IN VERSION 0.15.1
-------------------------
 o Proteins,EnsDb method allowing to fetch a Proteins object from an EnsDb
   database.
 o Fixes in the mapToGenome method:
   - Works also for negative strand encoded proteins (issue #29).
   - Supports a GRangesList object with arbitrary mcols.
   - Faster implementation (issue #30).
 o mapToGenome,Proteins,EnsDb and pmapToGenome,Proteins,EnsDb methods that
   map peptide features to the genome using annotations fetched from an EnsDb.
 o The seqnames of the Proteins object are used as names for the resulting
   GRangesList object from the mapToGenome and pmapToGenome methods (issue #34).

CHANGES IN VERSION 0.15.1
-------------------------
 o Drop unique `seqnames` requirement; see #28, #32
 o Create `names` as synonym for `seqnames`; close #32

CHANGES IN VERSION 0.13.3
-------------------------
 o Update mapping vignette index entry, as suggested by Mike Love
   <2016-09-13 Tue>

CHANGES IN VERSION 0.13.0
-------------------------
 o Bioc devel

CHANGES IN VERSION 0.11.3
-------------------------
 o acols and pranges replacement method <2016-02-22 Mon>

CHANGES IN VERSION 0.11.2
-------------------------
 o add addPeptideFragments; see #24 for details [2016-02-07 Sun]

CHANGES IN VERSION 0.11.1
-------------------------
 o elementLengths was renamed -> elementNROWS in S4Vectors (new name
   reflects TRUE semantic, old name will be deprecated soon) [r113044
   | hpages@fhcrc.org | 2016-01-29 01:22:03 +0000 (Fri, 29 Jan 2016)]

CHANGES IN VERSION 0.11.0
-------------------------
 o New Bioc devel version

CHANGES IN VERSION 0.10.0
-------------------------
 o New Bioc release version

CHANGES IN VERSION 0.9.1
------------------------
 o update cleave method to address build error [2015-10-13 Tue]

CHANGES IN VERSION 0.9.0
------------------------
 o Bioc devel 3.2

CHANGES IN VERSION 0.8.0
------------------------
 o Bioc release 3.1

CHANGES IN VERSION 0.6.13
-------------------------
 o Updating Biomart query to reflect changes [2015-04-06 Mon]

CHANGES IN VERSION 0.6.12
-------------------------
 o Fix filling in tracks plots [2015-03-07 Sat]

CHANGES IN VERSION 0.6.11
-------------------------
 o Documentation for the plotting heler functions [2015-02-28 Sat]

CHANGES IN VERSION 0.6.10
-------------------------
 o plotting helper functions [2015-02-28 Sat]
 o fixed peptide grouping [2015-02-28 Sat]

CHANGES IN VERSION 0.6.9
------------------------
 o Fixing junction peptides splitting to support peptides reaching > 2
   exons [2015-02-27 Fri]
 o supporting more protein coding biotypes [2015-02-27 Fri]
 o rewrite and simplify proteinCoverage; closes #19 [2015-02-27 Fri]

CHANGES IN VERSION 0.6.8
------------------------
 o fix splitting exon junction ranges [2015-02-26 Thu]

CHANGES IN VERSION 0.6.7
------------------------
 o The genomic coordinates of peptides that overlap exon junctions are
   now correctly split instead of spanning over the exon and are
   grouped accordingly. [2015-02-26 Thu]

CHANGES IN VERSION 0.6.6
------------------------
 o update mapping vignette and use alignment to identify correct
   transcript [2015-02-25 Wed]

CHANGES IN VERSION 0.6.5
------------------------
 o see also mapping vignette in ?mapToGenome [2015-02-20 Fri]
 o calculateHeavyLabels, isCleaved, proteotypic and proteinCoverage
   are is now a function [2015-02-20 Fri]
 o added accessors to Pbase:::pplot() and update Pbase-data vignette
   accordingly [2015-02-21 Sat]

CHANGES IN VERSION 0.6.4
------------------------
 o typos in man [2015-02-19 Thu]
 o Added use.names argument ot etrid2gel [2015-02-19 Thu]
 o mapToGenome,Proteins,GRangesList (remove GRanges) using names for
   1-to-many mapping and document in mapping vignette [2015-02-19 Thu]

CHANGES IN VERSION 0.6.3
------------------------
 o new fig in Pbase-data vignette [2015-02-14 Sat]
 o new section for pmapToGenome in mapping vignette [2015-02-14 Sat]

CHANGES IN VERSION 0.6.2
------------------------
 o Add uniprot release to p metadata [2015-02-14 Sat]
 o added a metadata update method [2015-02-14 Sat]

CHANGES IN VERSION 0.6.1
------------------------
 o [p]mapToGenome methods [2015-02-14 Sat]

CHANGES IN VERSION 0.6.0
------------------------
 o fixed typos in vigs [2015-02-12 Thu]
 o new etrid2grl function to convert Ensembl transcript identifiers to
   GRangesList objects [2015-02-13 Fri]
 o initial mapToGenome function [2015-02-13 Fri]
 o regenerated p data using latest UniProt release and added
   acols(p)$ENST [2015-02-13 Fri]

CHANGES IN VERSION 0.5.5
------------------------
 o various fixes in vignettes [2015-02-12 Thu]

CHANGES IN VERSION 0.5.4
------------------------
 o Ensembl - UCSC mapping vignette [2015-02-11 Wed]

CHANGES IN VERSION 0.5.3
------------------------
 o Render vignettes with rmarkdown [2015-02-11 Wed]
 o Improve mapping vignette [2015-02-11 Wed]

CHANGES IN VERSION 0.5.2
------------------------
 o Adapt to new `addIdentification` signature in MSnbase >= 1.15.5.
   [2015-02-10 Tue]

CHANGES IN VERSION 0.5.1
------------------------
 o Update vignette since TxDb's UCSC coordinates differ from those
   fetched from Ensembl via biomaRt. [2015-01-05 Mon]
 o Add `isCleaved` method. [2015-01-03 Sat]
 o Add `calculateHeavyLabels`; closes #6 and cleaver#5. [2015-01-03 Sat]

CHANGES IN VERSION 0.5.0
------------------------
 o Bump version for new Bioc release.

CHANGES IN VERSION 0.2.2

sandbox/code.R
library("Pbase")
f0 <- "~/Data2/Thermo_HELA_PRT/swissprot_human_canonical_19_09_12.fasta"
p0 <- Proteins(f0)

p1 <- addIdentificationData(p0, "~/Data2/Thermo_HELA_PRT/Thermo_Hela_PRTC_1.mzid",
                            TRUE)

p2 <- addIdentificationData(p0, "~/Data2/Thermo_HELA_PRT/Thermo_Hela_PRTC_1.mzid",
                            FALSE)

## only 10 proteins, chosen as follows
## lns <- elementNROWS(pranges(p0))
## plns <- elementNROWS(aa(p0))
## i <- which(lns > 20)
## j <- which(lns < 6 & lns > 3)
## c(i[1:5], j[1:5])

f <- "./swissprot_human_canonical_10.fasta"
p <- Proteins(f)
p <- addIdentificationData(p, "~/Data2/Thermo_HELA_PRT/Thermo_Hela_PRTC_1.mzid")

save(p, file = "../data/p.rda")

plot(p)


## at <- ATrack(start = c(250, 480), end = c(320, 520),
##              name = "Annotations")
## pat1 <- ProteinAxisTrack()
## pat2 <- ProteinAxisTrack()
## plotTracks(pat, from = 1, to = 850)

## pat <- ProteinAxisTrack(addNC = TRUE, littleTicks = TRUE)
## seqEx <- aa(p)[[1]]
## plotTracks(trackList = c(pat, st), from = 1, to = 40)


## pat<-ProteinAxisTrack(addNC = TRUE, littleTicks = TRUE)
## data(pep_hxb2)
## hxb2_seq <- metadata(pep_hxb2)$sequence
## st <- ProteinSequenceTrack(sequence = hxb2_seq, name = "env")
## s2 <- paste(strsplit(hxb2_seq, "")[[1]][1:100], collapse = "")
## st2 <- ProteinSequenceTrack(sequence = s2, name = "s2")
## plotTracks(trackList = c(pat, st, st2), from = 1, to = 500)

## png("p1.png")
## plotTracks(trackList = c(pat, st, st2), from = 1, to = 200)
## dev.off()

## png("p2.png")
## plotTracks(trackList = c(pat, st, st2), from = 1, to = 1000)
## dev.off()

library("pRoloc")
## load("~/ResearchProjects/collaborations/fusion-data-analysis/data/pdRes-2014-03-22.rda", verbose = TRUE)
load("~/ResearchProjects/collaborations/fusion-data-analysis/data/svmRes-2014-04-15-top.25.rda", verbose=TRUE)
setStockcol(NULL)
setStockcol(paste0(getStockcol(), "80"))

ptsze <- exp(fData(svmRes.2014.03.27)$svm.scores) - 1


pdf("../Figures/fus.pdf")
plot2D(svmRes.2014.03.27, fcol = "svm")
addLegend(svmRes.2014.03.27, fcol = "svm",
          where = "bottomleft", bty = "n",
          cex = .8)
dev.off()

########################################

library("biomaRt")

## Accession: P53501
## Name: ACT3_DROME
## FBgn0000044

## P60709 (ACTB_HUMAN) 
## ENSG00000075624

ens <- useMart("ensembl", "dmelanogaster_gene_ensembl")

select(ens, keys = "P53501",
       keytype = "uniprot_swissprot",
       columns = c(columns(ens)[1:12],
           "uniprot_swissprot"))

select(ens, keys = "ACT3_DROME",
       keytype = "uniprot_swissprot",
       columns = c(columns(ens)[1:12],
           "uniprot_swissprot"))

ens <- useMart("ensembl", "hsapiens_gene_ensembl")

select(ens, keys = "P60709",
       keytype = "uniprot_swissprot",
       columns = c(columns(ens)[1:12],
           "uniprot_swissprot"))

select(ens, keys = "ACTB_HUMAN",
       keytype = "uniprot_swissprot",
       columns = c(columns(ens)[1:12],
           "uniprot_swissprot"))



ens <- useMart("ensembl", "mmusculus_gene_ensembl")


bm <- select(ens, keys = "DTX4_MOUSE",
             keytype = "uniprot_swissprot",
             columns = c("ensembl_exon_id",
                 "peptide",
                 "exon_chrom_start",
                 "exon_chrom_end"))

bm <- select(ens, keys = "DTX4_MOUSE",
             keytype = "uniprot_swissprot",
             columns = c(
                 "ensembl_gene_id",
                 "ensembl_transcript_id",
                 "ensembl_peptide_id",
                 "ensembl_exon_id",  
                 "description",     
                 "chromosome_name",
                 "strand",
                 "exon_chrom_start",
                 "exon_chrom_end",
                 "5_utr_start",
                 "5_utr_end",
                 "3_utr_start",
                 "3_utr_end",
                 "start_position",     
                 "end_position"))

                 ## "band",
                 ## "transcript_start",
                 ## "transcript_end",
                 ## "exon_chrom_start",
                 ## "exon_chrom_end",
                 ## "rank",
                 ## "uniprot_swissprot",
                 ## "uniprot_swissprot",
                 ## "gene_exon",
                 ## "cdna",
                 ## "coding",
                 ## "peptide"))


library("AnnotationHub")
hub <- AnnotationHub()

## data exploration
length(names(hub))  ## resources available
md <- metadata(hub) ## DataFrame

dna <- hub$ensembl.release.74.fasta.mus_musculus.dna.Mus_musculus.GRCm38.74.dna.toplevel.fa.rz
## dnax <- readDNAStringSet(dna$index)
pep <- hub$ensembl.release.74.fasta.mus_musculus.pep.Mus_musculus.GRCm38.74.pep.all.fa.rz
gtf <- hub$ensembl.release.74.gtf.mus_musculus.Mus_musculus.GRCm38.74.gtf_0.0.1.RData

i <- which(gtf$protein_id == "ENSMUSP00000040229")
gtf[i]

library("Gviz")

library("BSgenome")
library("BSgenome.Mmusculus.UCSC.mm10")
mus <- Mmusculus
## getSeq(mus, gtf[i])

library(TxDb.Mmusculus.UCSC.mm10.knownGene)
xdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
##txdb <- extractTranscriptSeqs(mus, TxDb.Mmusculus.UCSC.mm10.knownGene)


####################
ens <- useMart("ensembl", "mmusculus_gene_ensembl")

## was not mapped by Alejandro
bm <- select(ens, keys = "SARNP_MOUSE",
             keytype = "uniprot_swissprot",
             columns = c(
                 "ensembl_gene_id",
                 "ensembl_transcript_id",
                 "ensembl_peptide_id",
                 "ensembl_exon_id",  
                 "description",     
                 "chromosome_name",
                 "strand",
                 "exon_chrom_start",
                 "exon_chrom_end",
                 "5_utr_start",
                 "5_utr_end",
                 "3_utr_start",
                 "3_utr_end",
                 "start_position",     
                 "end_position"))

###################################


library("BSgenome.Hsapiens.UCSC.hg19")
library("TxDb.Hsapiens.UCSC.hg19.knownGene")

txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

tx_seqs <- extractTranscriptSeqs(BSgenome.Hsapiens.UCSC.hg19, txdb)

tr <- transcripts(txdb)

(cds <- cdsBy(txdb))
cds_seqs <- extractTranscriptSeqs(BSgenome.Hsapiens.UCSC.hg19, cds)
